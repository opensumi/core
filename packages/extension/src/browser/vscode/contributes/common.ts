import { IExtension, JSONType } from '../../../common';
import { Disposable } from '@ide-framework/ide-core-common';

export const CONTRIBUTE_NAME_KEY = 'contribute_name';
export abstract class VscodeContributionPoint<T extends JSONType = JSONType> extends Disposable {

  constructor(protected json: T, protected contributes: any, protected extension: IExtension) {
    super();
  }

  abstract contribute();

}

export function Contributes(name) {
  return (target) => {
    Reflect.defineMetadata(CONTRIBUTE_NAME_KEY, name, target);
  };
}
