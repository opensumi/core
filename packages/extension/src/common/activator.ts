import { IDisposable } from '@opensumi/ide-core-common';

import { ExtensionHostType } from '.';

import type vscode from 'vscode';

export class ExtensionActivationTimes {
  public static readonly NONE = new ExtensionActivationTimes(false, -1, -1, -1);

  constructor(
    public readonly startup: boolean,
    public readonly codeLoadingTime: number,
    public readonly activateCallTime: number,
    public readonly activateResolvedTime: number,
  ) {}
}

export interface IExtensionModule {
  activate?(ctx: vscode.ExtensionContext): Promise<any>;
  deactivate?(): void;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface IExtensionAPI {}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface IExtendExportAPI {}

export interface ActivatedExtensionJSON {
  id: string;
  host: ExtensionHostType;
  activationFailed: boolean;
  activationFailedError: Error | null;
  activateCallTime?: number;
}

export class ActivatedExtension {
  constructor(
    public readonly id: string,
    public readonly displayName: string,
    public readonly description: string,
    public readonly host: ExtensionHostType,
    public readonly activationFailed: boolean,
    public readonly activationFailedError: Error | null,
    public readonly module: IExtensionModule,
    public readonly exports: IExtensionAPI | undefined,
    public readonly subscriptions: IDisposable[],
    public readonly activationTimes?: ExtensionActivationTimes,
    public readonly extendExports?: IExtendExportAPI,
    public readonly extendModule?: IExtensionModule,
  ) {}

  public toJSON() {
    return {
      id: this.id,
      displayName: this.displayName,
      description: this.description,
      host: this.host,
      activationFailed: this.activationFailed,
      activationFailedError: this.activationFailedError,
      activateCallTime: this.activationTimes?.activateCallTime,
    };
  }
}

export class ExtensionsActivator {
  private readonly activatedExtensions: Map<string, ActivatedExtension> = new Map();

  constructor(private logger: any = console) {}

  has(id: string) {
    return this.activatedExtensions.has(id);
  }

  set(id: string, extension: ActivatedExtension) {
    return this.activatedExtensions.set(id, extension);
  }

  get(id: string): ActivatedExtension | undefined {
    return this.activatedExtensions.get(id);
  }

  all(): ActivatedExtension[] {
    return Array.from(this.activatedExtensions.values());
  }

  delete(id: string) {
    return this.activatedExtensions.delete(id);
  }

  private async doDeactivate(extensionModule) {
    try {
      const promiseLike = await (extensionModule as Required<IExtensionModule>).deactivate();
      return promiseLike;
    } catch (err) {
      this.logger.error(`
        [Extension-Host] deactivate extension module error ${err.message} \n\n
        Stack: ${err.stack && err.stack}
      `);
      return err;
    }
  }

  deactivate() {
    const deactivateTasks: Promise<void>[] = [];
    this.activatedExtensions.forEach((ext) => {
      const extModule = ext.module;
      if (extModule && extModule.deactivate && typeof extModule.deactivate === 'function') {
        deactivateTasks.push(this.doDeactivate(extModule));
      }

      const extendModule = ext.extendModule;
      if (extendModule && extendModule.deactivate && typeof extendModule.deactivate === 'function') {
        deactivateTasks.push(this.doDeactivate(extendModule));
      }

      ext.subscriptions.forEach((disposable) => {
        try {
          disposable.dispose();
        } catch (e) {
          this.logger.log('extension deactivated error');
          this.logger.warn(e);
        }
      });
    });
    return Promise.all(deactivateTasks);
  }
}
