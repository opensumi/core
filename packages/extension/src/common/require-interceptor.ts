import { Injectable, Injector, Autowired } from '@opensumi/di';
import { IRPCProtocol } from '@opensumi/ide-connection/lib/common/rpcProtocol';
import { Disposable, IDisposable, ILogger } from '@opensumi/ide-core-common';
import { IExtension } from '..';

export interface IBrowserRequireInterceptorArgs {
  injector: Injector;
  extension: IExtension;
  rpcProtocol?: IRPCProtocol;
}

export interface IRequireInterceptor<T> {
  moduleName: string;
  load(request: T): any;
}

export const IRequireInterceptorService = Symbol('IRequireInterceptorService');
export interface IRequireInterceptorService<T = any> {
  registerRequireInterceptor(interceptor: IRequireInterceptor<T>): IDisposable;
  getRequireInterceptor(moduleName: string): IRequireInterceptor<T> | undefined;
}

export const RequireInterceptorContribution = Symbol('RequireInterceptorContribution');

export interface RequireInterceptorContribution {
  registerRequireInterceptor(registry: IRequireInterceptorService): void;
}

@Injectable()
export class RequireInterceptorService<T> implements IRequireInterceptorService<T> {

  @Autowired(ILogger)
  private readonly logger: ILogger;

  private readonly registry = new Map<string, IRequireInterceptor<T>>();

  registerRequireInterceptor(interceptor: IRequireInterceptor<T>) {
    if (this.registry.has(interceptor.moduleName)) {
      this.logger.warn(`module ${interceptor.moduleName} already register`);
      return Disposable.NULL;
    }
    this.registry.set(interceptor.moduleName, interceptor);
    return Disposable.create(() => {
      this.registry.delete(interceptor.moduleName);
    });
  }

  getRequireInterceptor(moduleName: string) {
    return this.registry.get(moduleName);
  }
}
