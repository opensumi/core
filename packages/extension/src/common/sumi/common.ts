
export interface IMainThreadCommon {
  $subscribeEvent(eventName): Promise<void>;
  $unSubscribeEvent(eventName): Promise<void>;
}

export interface IExtHostCommon {
  $acceptEvent(eventName: string, eventArgs: any[]): Promise<any[]>;
}

export interface IExtHostEventResult {
  err?: string;
  result: any;
}
