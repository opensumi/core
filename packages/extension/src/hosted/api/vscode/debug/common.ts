import type vscode from 'vscode';

import { DebugConfigurationProviderTriggerKind } from '../../worker/worker.ext-types';

export interface IDebugConfigurationProvider extends vscode.DebugConfigurationProvider {
  readonly type: string;
  readonly triggerKind: DebugConfigurationProviderTriggerKind;
}
