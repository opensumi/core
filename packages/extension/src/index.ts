export * from './common';
export * from './node';
