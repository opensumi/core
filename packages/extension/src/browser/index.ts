import { Injectable, Provider } from '@opensumi/di';
import { BrowserModule } from '@opensumi/ide-core-browser';
import { DebugSessionContributionRegistry } from '@opensumi/ide-debug/lib/browser/debug-session-contribution';
import { IDebugServer } from '@opensumi/ide-debug/lib/common';
import { FileSearchServicePath } from '@opensumi/ide-file-search/lib/common';

import {
  AbstractExtensionManagementService,
  ExtensionHostProfilerServicePath,
  ExtensionNodeServiceServerPath,
  ExtensionService,
  IExtCommandManagement,
  IRequireInterceptorService,
  RequireInterceptorContribution,
  RequireInterceptorService,
} from '../common';
import {
  AbstractNodeExtProcessService,
  AbstractViewExtProcessService,
  AbstractWorkerExtProcessService,
} from '../common/extension.service';
import {
  IMainThreadExtenderService,
  MainThreadExtenderContribution,
  MainThreadExtenderService,
} from '../common/main.thread.extender';

import { ActivationEventServiceImpl } from './activation.service';
import { ExtCommandManagementImpl as ExtCommandManagementImpl } from './extension-command-management';
import { ExtInstanceManagementService } from './extension-instance-management';
import { ExtensionManagementService } from './extension-management.service';
import { NodeExtProcessService } from './extension-node.service';
import { ViewExtProcessService } from './extension-view.service';
import { WorkerExtProcessService } from './extension-worker.service';
import { ExtensionClientAppContribution, ExtensionCommandContribution } from './extension.contribution';
import { ExtensionServiceImpl } from './extension.service';
import { BrowserRequireInterceptorContribution } from './require-interceptor.contribution';
import { SumiContributionsService, SumiContributionsServiceToken } from './sumi/contributes';
import { AbstractExtInstanceManagementService, IActivationEventService } from './types';
import { ExtensionDebugService, ExtensionDebugSessionContributionRegistry } from './vscode/api/debug';
import { VSCodeContributesService, VSCodeContributesServiceToken } from './vscode/contributes';

@Injectable()
export class ExtensionModule extends BrowserModule {
  contributionProvider = [RequireInterceptorContribution, MainThreadExtenderContribution];
  providers: Provider[] = [
    {
      token: ExtensionService,
      useClass: ExtensionServiceImpl,
    },
    {
      token: IDebugServer,
      useClass: ExtensionDebugService,
      override: true,
    },
    {
      token: DebugSessionContributionRegistry,
      useClass: ExtensionDebugSessionContributionRegistry,
      override: true,
    },
    {
      token: IActivationEventService,
      useClass: ActivationEventServiceImpl,
    },
    {
      token: IExtCommandManagement,
      useClass: ExtCommandManagementImpl,
    },
    {
      token: AbstractExtInstanceManagementService,
      useClass: ExtInstanceManagementService,
    },
    {
      token: AbstractExtensionManagementService,
      useClass: ExtensionManagementService,
    },
    {
      token: AbstractNodeExtProcessService,
      useClass: NodeExtProcessService,
    },
    {
      token: AbstractWorkerExtProcessService,
      useClass: WorkerExtProcessService,
    },
    {
      token: AbstractViewExtProcessService,
      useClass: ViewExtProcessService,
    },
    {
      token: IRequireInterceptorService,
      useClass: RequireInterceptorService,
    },
    {
      token: VSCodeContributesServiceToken,
      useClass: VSCodeContributesService,
    },
    {
      token: SumiContributionsServiceToken,
      useClass: SumiContributionsService,
    },
    {
      token: IMainThreadExtenderService,
      useClass: MainThreadExtenderService,
    },
    ExtensionCommandContribution,
    ExtensionClientAppContribution,
    BrowserRequireInterceptorContribution,
  ];

  backServices = [
    {
      servicePath: ExtensionNodeServiceServerPath,
      clientToken: ExtensionService,
    },
    {
      servicePath: FileSearchServicePath,
    },
    {
      servicePath: ExtensionHostProfilerServicePath,
    },
  ];
}
