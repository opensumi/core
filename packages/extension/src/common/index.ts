import {
  Disposable,
  IJSONSchema,
  IDisposable,
  Deferred,
  Uri,
  MaybePromise,
  IExtensionLogger,
  ExtensionConnectOption,
  replaceNlsField,
  ILogger,
} from '@opensumi/ide-core-common';
import { createExtHostContextProxyIdentifier } from '@opensumi/ide-connection';
import { ExtHostStorage } from '../hosted/api/vscode/ext.host.storage';
import { Extension } from '../hosted/vscode.extension';
import { Emitter, IExtensionProps } from '@opensumi/ide-core-common';
import { IExtensionContributions, IMainThreadCommands } from './vscode';
import { ISumiExtensionContributions } from './sumi/extension';
import { ActivatedExtension, ExtensionsActivator, ActivatedExtensionJSON } from './activator';
import { typeAndModifierIdPattern } from '@opensumi/ide-theme/lib/common/semantic-tokens-registry';

export { IExtensionProps } from '@opensumi/ide-core-common';

export * from './ext.host.proxy';
export * from './require-interceptor';

export interface IExtensionMetaData {
  id: string;
  extensionId: string;
  // 支持使用自定义uri
  path: string;
  uri?: Uri;
  packageJSON: { [key: string]: any };
  defaultPkgNlsJSON: { [key: string]: any } | undefined;
  packageNlsJSON: { [key: string]: any } | undefined;
  extraMetadata: JSONType;
  realPath: string; // 真实路径，用于去除symbolicLink
  extendConfig: JSONType;
  isBuiltin: boolean;
  isDevelopment?: boolean;
}

/**
 * 提供插件扫描时的额外信息读取能力
 * 比如 read/changelog/package.nls.json 等等
 */
export interface IExtraMetaData {
  [key: string]: string;
}

export const ExtensionNodeServiceServerPath = 'ExtensionNodeServiceServerPath';

export type ExtensionDependencies = (string | { [extensionId: string]: string })[];

export interface ICreateProcessOptions {
  /**
   * 启用插件进程的 Debug 模式
   */
  enableDebugExtensionHost?: boolean;
  /**
   * 插件进程连接时候一些配置选项
   */
  extensionConnectOption?: ExtensionConnectOption;
}

export const IExtensionNodeService = Symbol('IExtensionNodeService');
export interface IExtensionNodeService {
  initialize(): Promise<void>;
  getAllExtensions(
    scan: string[],
    extensionCandidate: string[],
    localization: string,
    extraMetaData: IExtraMetaData,
  ): Promise<IExtensionMetaData[]>;
  createProcess(clientId: string, options?: ICreateProcessOptions): Promise<void>;
  ensureProcessReady(clientId: string): Promise<boolean>;
  getElectronMainThreadListenPath(clientId: string);
  getElectronMainThreadListenPath2(clientId: string);
  getExtServerListenOption(clientId: string);
  getExtension(
    extensionPath: string,
    localization: string,
    extraMetaData?: IExtraMetaData,
  ): Promise<IExtensionMetaData | undefined>;
  setConnectionServiceClient(clientId: string, serviceClient: IExtensionNodeClientService);
  disposeClientExtProcess(clientId: string, info: boolean): Promise<void>;
  disposeAllClientExtProcess(): Promise<void>;
  tryEnableInspectPort(clientId: string, delay?: number): Promise<boolean>;
  getProcessInspectPort(clientId: string): Promise<number | undefined>;
}

export const IExtensionNodeClientService = Symbol('IExtensionNodeClientService');
export interface IExtensionNodeClientService {
  getElectronMainThreadListenPath(clientId: string): Promise<string>;
  getAllExtensions(
    scan: string[],
    extensionCandidate: string[],
    localization: string,
    extraMetaData: IExtraMetaData,
  ): Promise<IExtensionMetaData[]>;
  createProcess(clientId: string, options: ICreateProcessOptions): Promise<void>;
  getExtension(
    extensionPath: string,
    localization: string,
    extraMetaData?: IExtraMetaData,
  ): Promise<IExtensionMetaData | undefined>;
  infoProcessNotExist(): void;
  infoProcessCrash(): void;
  restartExtProcessByClient(): void;
  disposeClientExtProcess(clientId: string, info: boolean): Promise<void>;
  updateLanguagePack(languageId: string, languagePackPath: string, storagePath: string): Promise<void>;
}

export type ExtensionHostType = 'node' | 'worker';

export type ExtensionHostTypeUpperCase = 'Node.js' | 'Web Worker';

export interface ChangeExtensionOptions {
  upgrade: boolean;
  extensionPath: string;
  oldExtensionPath?: string;
  isBuiltin?: boolean;
}

/**
 * 管理不同进程内插件注册的 command 的运行环境及其调用
 */
export abstract class IExtCommandManagement {
  abstract registerProxyCommandExecutor(env: ExtensionHostType, proxyCommandExecutor: IMainThreadCommands): void;
  abstract executeExtensionCommand(env: ExtensionHostType, command: string, args: any[]): Promise<any>;
  /**
   * @param command command id
   * @param targetHost 目标插件进程的运行环境，默认 'node'
   */
  abstract registerExtensionCommandEnv(command: string, targetHost?: ExtensionHostType): IDisposable;
  abstract getExtensionCommandEnv(command: string): ExtensionHostType | undefined;
}

/**
 * 为插件市场面板提供数据/交互
 */
export abstract class AbstractExtensionManagementService {
  /**
   * @deprecated 建议直接用 AbstractExtInstanceManagementService#getExtensionInstances 后自行 map#toJSON 即可
   */
  abstract getAllExtensionJson(): IExtensionProps[];

  /**
   * 禁用插件
   */
  abstract postDisableExtension(extensionPath: string): Promise<void>;

  /**
   * 启用插件
   */
  abstract postEnableExtension(extensionPath: string): Promise<void>;

  /**
   * 安装插件之后开始激活
   */
  abstract postChangedExtension(options: ChangeExtensionOptions): Promise<void>;
  abstract postChangedExtension(upgrade: boolean, extensionPath: string, oldExtensionPath?: string): Promise<void>;
  abstract postChangedExtension(
    upgrade: boolean | ChangeExtensionOptions,
    extensionPath?: string,
    oldExtensionPath?: string,
  ): Promise<void>;

  /**
   * 卸载插件
   */
  abstract postUninstallExtension(extensionPath: string): Promise<void>;

  /**
   * 通过 extensionPath 获取插件实例
   */
  abstract getExtensionByPath(extensionPath: string): IExtension | undefined;
  /**
   * 通过 extension id 获取插件实例
   */
  abstract getExtensionByExtId(extensionId: string): IExtension | undefined;

  /**
   * 通过 extensionPath 获取插件实例序列化数据及从 node 层获取的 extraMetadata
   */
  abstract getExtensionProps(
    extensionPath: string,
    extraMetaData?: IExtraMetaData,
  ): Promise<IExtensionProps | undefined>;
}

export abstract class ExtensionService {
  /**
   * 激活插件服务
   */
  abstract activate(): Promise<void>;

  /**
   * 重启插件进程
   */
  abstract restartExtProcess(): Promise<void>;

  /**
   * 激活插件, 给 Extension 实例使用
   */
  abstract activeExtension(extension: IExtension): Promise<void>;

  /**
   * 销毁插件
   */
  abstract disposeExtensions(): Promise<void>;

  /**
   * 执行插件命令
   */
  abstract executeExtensionCommand(command: string, args: any[]): Promise<void>;

  /**
   * @internal 提供获取所有运行中的插件的列表数据
   */
  abstract getActivatedExtensions(): Promise<{ [key in ExtensionHostType]?: ActivatedExtension[] }>;

  eagerExtensionsActivated: Deferred<void>;
}

export abstract class ExtensionCapabilityRegistry {
  abstract getAllExtensions(): Promise<IExtensionMetaData[]>;
}

export const LANGUAGE_BUNDLE_FIELD = 'languageBundle';

export interface JSONType {
  [key: string]: any;
}

export interface IExtension extends IExtensionProps {
  readonly contributes: IExtensionContributions & ISumiExtensionContributions;
  activate(visited?: Set<string>);
  enable(): void;
  reset(): void;
  toJSON(): IExtensionProps;
}

//  VSCode Types
export abstract class VSCodeContributePoint<T extends JSONType = JSONType> extends Disposable {
  constructor(
    protected json: T,
    protected contributes: any,
    protected extension: IExtensionMetaData,
    protected packageNlsJSON: JSONType | undefined,
    protected defaultPkgNlsJSON: JSONType | undefined,
  ) {
    super();
  }
  schema?: IJSONSchema;

  abstract contribute();

  protected getLocalizeFromNlsJSON(title: string): string {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return replaceNlsField(title, this.extension.id)!;
  }
}

export const CONTRIBUTE_NAME_KEY = 'contribute_name';
export function Contributes(name) {
  return (target) => {
    Reflect.defineMetadata(CONTRIBUTE_NAME_KEY, name, target);
  };
}

export const EXTENSION_EXTEND_SERVICE_PREFIX = 'extension_extend_service';
export const MOCK_EXTENSION_EXTEND_PROXY_IDENTIFIER = createExtHostContextProxyIdentifier(
  'mock_extension_extend_proxy_identifier',
);

export interface IExtensionHost {
  logger: IExtensionLogger;
  $activateExtension(id: string): Promise<void>;
  /**
   * 更新插件进程内的插件相关数据
   */
  $updateExtHostData(): Promise<void>;
  $getActivatedExtensions(): Promise<ActivatedExtensionJSON[]>;

  getExtensionExports(id: string): any;
  getExtensions(): Extension[];
  getExtension(extensionId: string): Extension<any, IExtensionHost> | undefined;
  isActivated(id: string): boolean;
  activateExtension(id: string): Promise<void>;
  extensionsChangeEmitter: Emitter<void>;
  storage: ExtHostStorage;
}

export interface IExtensionHostService extends IExtensionHost {
  $fireChangeEvent(): Promise<void>;
  init(): Promise<void>;
  close(): Promise<void>;
  getExtendExports(id: string): any;
  extensionsActivator: ExtensionsActivator;
  /**
   * 上报插件未捕获异常
   * @param error
   */
  reportUnexpectedError(error: Error): void;
}

export interface IExtensionWorkerHost extends IExtensionHost {
  staticServicePath: string;
}

export interface IExtendProxy {
  [key: string]: any;
}

export const WorkerHostAPIIdentifier = {
  ExtWorkerHostExtensionService: createExtHostContextProxyIdentifier('ExtWorkerHostExtensionService'),
};

export enum ProcessMessageType {
  REPORTER,
}

export enum EXTENSION_ENABLE {
  ENABLE = 1,
  DISABLE = 0,
}

// 广播插件事件
export const EMIT_EXT_HOST_EVENT = {
  id: 'sumi-extension:ext-host-event',
};

export function getExtensionId(extensionId: string) {
  return extensionId.toLowerCase();
}

export enum ExtensionHostKind {
  NODE_HOST = 1,
  WORKER_HOST = 2,
}

export const ExtensionHostProfilerServicePath = 'ExtensionHostProfilerService';

export const ExtensionHostProfilerServiceToken = Symbol('ExtensionHostProfilerService');

export interface IExtensionHostProfilerService {
  $startProfile(clientId: string): Promise<void>;
  $stopProfile(clientId: string): Promise<boolean>;
  $saveLastProfile(filePath: string): Promise<void>;
}

export enum OutputType {
  STDOUT,
  STDERR,
}

export interface Output {
  type: OutputType;
  data: string;
  format: string[];
}

export const IExtensionHostManager = Symbol('IExtensionHostManager');

export interface IExtensionHostManager {
  init(): MaybePromise<void>;
  fork(modulePath: string, ...args: any[]): MaybePromise<number>;
  send(pid: number, message: string): MaybePromise<void>;
  isRunning(pid: number): MaybePromise<boolean>;
  treeKill(pid: number): MaybePromise<void>;
  kill(pid: number, signal?: string): MaybePromise<void>;
  isKilled(pid: number): MaybePromise<boolean>;
  findDebugPort(startPort: number, giveUpAfter: number, timeout: number): Promise<number>;
  onOutput(pid: number, listener: (output: Output) => void): MaybePromise<void>;
  onExit(pid: number, listener: (code: number, signal: string) => void): MaybePromise<void>;
  onMessage(pid: number, listener: (msg: any) => void): MaybePromise<void>;
  disposeProcess(pid: number): MaybePromise<void>;
  dispose(): MaybePromise<void>;
}

/**
 * 以下都是 fork 新的插件子进程的时候，传入的命令行参数的 key
 */
export const KT_PROCESS_SOCK_OPTION_KEY = 'kt-process-sock-option';
export const KT_PROCESS_PRELOAD_KEY = 'kt-process-preload';
export const KT_APP_CONFIG_KEY = 'kt-app-config';

// #region Semantic Tokens Contribution Point

export interface SemanticTokenScopes {
  scopes?: { [selector: string]: string[] };
  language?: string;
}

export type SemanticTokenScopesSchema = Array<SemanticTokenScopes>;

export interface SemanticTokenType {
  id: string;
  description: string;
  superType?: string;
}

export type SemanticTokenTypeSchema = Array<SemanticTokenType>;

export interface SemanticTokenModifier {
  id: string;
  description: string;
}

export type SemanticTokenModifierSchema = Array<SemanticTokenModifier>;

export function validateTypeOrModifier(
  contribution: SemanticTokenType | SemanticTokenModifier,
  extensionPoint: string,
  logger: ILogger,
): boolean {
  if (typeof contribution.id !== 'string' || contribution.id.length === 0) {
    logger.error("'configuration.{0}.id' must be defined and can not be empty", extensionPoint);
    return false;
  }
  if (!contribution.id.match(typeAndModifierIdPattern)) {
    logger.error("'configuration.{0}.id' must follow the pattern letterOrDigit[-_letterOrDigit]*");
    return false;
  }
  const superType = (contribution as SemanticTokenType).superType;
  if (superType && !superType.match(typeAndModifierIdPattern)) {
    logger.error(
      "'configuration.{0}.superType' must follow the pattern letterOrDigit[-_letterOrDigit]*",
      extensionPoint,
    );
    return false;
  }
  if (typeof contribution.description !== 'string' || contribution.id.length === 0) {
    logger.error("'configuration.{0}.description' must be defined and can not be empty", extensionPoint);
    return false;
  }
  return true;
}

// #endregion Semantic Tokens Contribution Point
