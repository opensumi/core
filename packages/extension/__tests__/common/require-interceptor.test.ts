import React from 'react';

import { createBrowserInjector } from '../../../../tools/dev-tool/src/injector-helper';
import { IRequireInterceptorService, RequireInterceptorService } from '../../src/common/require-interceptor';

describe(__filename, () => {
  const injector = createBrowserInjector([]);
  let requireInterceptorService: IRequireInterceptorService;

  beforeEach(() => {
    injector.addProviders({
      token: IRequireInterceptorService,
      useClass: RequireInterceptorService,
    });
    requireInterceptorService = injector.get(IRequireInterceptorService);
  });

  afterEach(() => {
    injector.disposeAll();
  });

  it('registerRequireInterceptor', () => {
    requireInterceptorService.registerRequireInterceptor({
      moduleName: 'react',
      load: () => React,
    });
    const interceptor = requireInterceptorService.getRequireInterceptor('react');
    const request = interceptor?.load({});
    expect(request).toBe(React);
  });
});
