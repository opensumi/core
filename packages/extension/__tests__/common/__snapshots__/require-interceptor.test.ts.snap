// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`require-interceptor test registerRequireInterceptor 1`] = `
{
  "Children": {
    "count": [Function],
    "forEach": [Function],
    "map": [Function],
    "only": [Function],
    "toArray": [Function],
  },
  "Component": [Function],
  "Fragment": Symbol(react.fragment),
  "Profiler": Symbol(react.profiler),
  "PureComponent": [Function],
  "StrictMode": Symbol(react.strict_mode),
  "Suspense": Symbol(react.suspense),
  "__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED": {
    "ReactCurrentActQueue": {
      "current": null,
      "didScheduleLegacyUpdate": false,
      "isBatchingLegacy": false,
    },
    "ReactCurrentBatchConfig": {
      "transition": null,
    },
    "ReactCurrentDispatcher": {
      "current": null,
    },
    "ReactCurrentOwner": {
      "current": null,
    },
    "ReactDebugCurrentFrame": {
      "getCurrentStack": null,
      "getStackAddendum": [Function],
      "setExtraStackFrame": [Function],
    },
  },
  "cloneElement": [Function],
  "createContext": [Function],
  "createElement": [Function],
  "createFactory": [Function],
  "createRef": [Function],
  "forwardRef": [Function],
  "isValidElement": [Function],
  "lazy": [Function],
  "memo": [Function],
  "startTransition": [Function],
  "unstable_act": [Function],
  "useCallback": [Function],
  "useContext": [Function],
  "useDebugValue": [Function],
  "useDeferredValue": [Function],
  "useEffect": [Function],
  "useId": [Function],
  "useImperativeHandle": [Function],
  "useInsertionEffect": [Function],
  "useLayoutEffect": [Function],
  "useMemo": [Function],
  "useReducer": [Function],
  "useRef": [Function],
  "useState": [Function],
  "useSyncExternalStore": [Function],
  "useTransition": [Function],
  "version": "18.2.0",
}
`;
